### ISSUE TITLE
missing return value into socket_cmp_addresses

### ISSUE THREAD

**Issue by chaostheory88** [tass-belgium/picotcp#390]
socket_cmp_addresses into pico_socket.c has missing return value, it returns always 0.

**Comment by danielinux** [comment]
Hello! Any news on this one?

**Comment by toonst** [comment]
Fixed in 07085fa62b5d611d98a24648af2d071990cb16ea

**Comment by frederikvs** [comment]
Thanks, @toonst, but it looks like this fix broke some stuff for a.o. @matthiasvp, related to listening sockets. Also, some of our Jenkins tests started failing. 

Thanks for reporting this, @chaostheory88, but can you let us know what problem you encountered? I'd like to add whatever problem you encountered to our tests, so we can look for a better fix, and be sure that it fixes your problem.

**Comment by toonst** [comment]
@frederikvs The implementation was broken before, so that probably means the tests are faulty. @matthiasvp can you specify which problems you have now that the socket_cmp_addresses actually does something? Maybe some bug in `socket_cmp_ipv6`, `socket_cmp_ipv4`, or `socket_cmp_remotehost` that was masked before?

**Comment by matthiasvp** [comment]
@toonst When running a tcp server the connection gets reset immediately after a client connects.
Some more investigation shows that after the socket cloned itself(when the client has connected) the original listening socket is still used instead of the newly cloned one.
This results in an invalid state triggering a reset.

To reproduce, simply create a TCP listener socket and connect to it.

So I guess there is indeed a bug in at least one of the functions you suggested.

**Comment by danielinux** [comment]
@frederikvs : @chaostheory88 did not really encounter a problem, he just reported that the function is not returning anything useful by looking at it. Following the logic of the module, it seems that `socket_cmp_address()` should in fact compare the address of the incoming frame to that on the socket endpoint, to see if there is a match.

Perhaps not fixing this has other side effects (e.g. accepting alien packets on sockets when the address don't match), but we haven't run any test on that yet. 

If this is not the case, then you would agree that this function is useless and should be removed as it always return the same value, regardless of the input.

**Comment by frederikvs** [comment]
Finally found some time to investigate further!

The function is not useless. If we would remove the function (or before @toonst's commit, when it always returned 0), `socket_cmp` would check IP version, and then remote port. Local port was already compared earlier. So, when by coincidence we would have two sockets with same local port and remote port, it wouldn't distinguish between them, even though remote IP was different.

Potentially a quite annoying issue indeed. Glad it got reported :-)

@toonst's fix looks correct at first glance, but still it breaks listening sockets somehow. And it seems to be related to the ordering of sockets in the tree. As an experiment, if I do return `-ret`, listening sockets work again. The order in the tree shouldn't really matter, but it does.
So it looks like at some point there's something walking our tree, not doing enough of a compare, and just picking the first socket with certain characteristics. If it's in one order, it happens to be the correct socket, if it's in the other order, it's not, and stuff breaks.
I didn't get quite to the bottom yet, but it's progress :-)


We also need to make sure the original issue doesn't bite us again. 
@maartenvds @ISY-Thomas do you think this is something we could do in robot framework? 
I think we need to open a listening socket, have two other "devices" connect to it, "coincidentally" choosing the same port. Then see if picoTCP can handle this.
(Note : if we create this test, we need to verify that it fails before @toonst's commit)

**Comment by frederikvs** [comment]
bit more time, bit more progress :

`pico_socket_deliver` (pico_scoket.c) calls `pico_socket_tcp_deliver` with the sockport and the frame. (The sockport is basically a tree of all sockets that share a protocol (e.g. TCP) and a local port.)
Then `pico_socket_tcp_deliver` loops over all sockets in that sockport tree, sees if they match (using `socket_tcp_deliver_ipv4`, or `6`).  When it finds a matching socket, it breaks out of the loop, and calls `socket_tcp_do_deliver` on that socket.

Now, if `socket_tcp_deliver_ipv4` would give a match with two sockets, we would only try to deliver it to the first we found. This would depend on the ordering in the sockport, and as I mentioned before, @toonst's fix changes the ordering of the sockets.

When we have both a listening socket (only the local port is known [*] ), and a connected socket (local port, remote port, remote IP are fixed), and a frame comes in, then both these sockets would match.
Seems to me that if a connected socket matches, we should deliver to that one, and only if there's no matching connected socket, we should go with the listening socket. 

[*] Potentially the local IP is known as well, depending on whether the user wanted to listen on a specific IP or on PICO_IPV4_INADDR_ANY, but it doesn't make a real difference, so lets just assume we're working with PICO_IPV4_INADDR_ANY here.



Seems to me that the simplest fix is to modify the looping over the sockets : only break out of the loop if the matching socket is connected. (If we looped around and we've not found a connected socket, we'll still go with the listening one)
Not yet entirely convinced that this will cover everything though, I'm open to suggestions/discussion/feedback. 


Note to future self : don't forget to fix both the IPv4 and IPv6 versions. And check if something like this exists for UDP as well.

**Comment by frederikvs** [comment]
(oops, accidentally pressed "close and comment" instead of "comment")

**Comment by frederikvs** [comment]
Out of curiosity, I've tracked down the history of these problems. The reported problem of `return 0;` was introduced in February 2014 (2ce9331c999d392558f830ec1fbc6d295284f42f). The problem of finding the right socket was introduced in April 2014 (b2ef46611bef1a96ff8c5c7d4bc0b964446e34f5).
The former issue hid the latter, while the latter is actually more severe.

Both commits were refactorings aimed at decreasing code complexity. I'm not saying it's a trend, it's just two datapoints, but still.

Anyway, the former was already fixed earlier, I think I've fixed the latter.
Unit and smoke tests pass locally, jenkins should start running all the other tests any moment now.

I think this can be closed.
Again, thanks for reporting @chaostheory88, and thanks everyone for the input and/or partial fix :-)
